### PR TITLE
[spirv] Translation of do-while loops

### PR DESCRIPTION
**Pull request by ehsannas**

### CODE REVIEW

**Comment by ehsannas** [comment]
Rebased.

File: tools/clang/include/clang/SPIRV/ModuleBuilder.h [hunk 1]
```diff
@@ -150,23 +150,39 @@ class ModuleBuilder {
...
           spv::SelectionControlMask::MaskNone,
       spv::LoopControlMask loopControl = spv::LoopControlMask::MaskNone);
 
+  /// \brief Creates a conditional branch (OpBranchConditional) without creating
+  /// an OpSelectionMerge!
+  /// **Note**
+  /// In SPIR-V, if a branch occurs in a continue block of a loop,
+  /// OpSelectionMerge should not be specified. In all other cases (all usual
+  /// cases), the 'createConditionalBranch' API should be used.
+  void createConditionalBranchWithoutSelectionMerge(uint32_t condition,
```
Comment (antiagainst) [inline tools/clang/include/clang/SPIRV/ModuleBuilder.h]: I think this case is already covered by the above `createConditionalBranch` method with all the parameters starting from `mergeLabel` set to their defaults?

File: tools/clang/include/clang/SPIRV/ModuleBuilder.h [hunk 2]
```diff
@@ -150,23 +150,39 @@ class ModuleBuilder {
...
   void createBranch(uint32_t targetLabel);
 
-  // \brief Creates a conditional branch. An OpSelectionMerge instruction
-  // will be created if mergeLabel is not 0 and continueLabel is 0.
-  // An OpLoopMerge instruction will also be created if both continueLabel
-  // and mergeLabel are not 0. For other cases, mergeLabel and continueLabel
-  // will be ignored. If selection control mask and/or loop control mask are
-  // provided, they will be applied to the corresponding SPIR-V instruction.
-  // Otherwise, MaskNone will be used.
+  /// \brief Creates an OpLoopMerge instruction followed by an uncondtiional
```
Comment (antiagainst) [inline tools/clang/include/clang/SPIRV/ModuleBuilder.h]: I think we can merge this overload with the above by giving `mergeBB` and `continueBB` defaults (0)?

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 3]
```diff
@@ -466,6 +468,96 @@ spv::LoopControlMask SPIRVEmitter::translateLoopAttribute(const Attr &attr) {
...
+  // Since SPIR-V requires structured control flow, we need two more basic
+  // blocks, <header> and <merge>. <header> is the block before control flow
+  // diverges, and <merge> is the block where control flow subsequently
+  // converges. The <check> can be performed in the <continue> basic block.
+  // The final CFG should normally be like the following. Exceptions
+  // will occur with non-local exits like loop breaks or early returns.
+  //
+  //            +----------+
+  //            |  header  | <-----------------------------------+
+  //            |  (true)  |                                     |
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: What's the purpose of this `(true)` part here? Seems can be removed.

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 4]
```diff
@@ -466,6 +468,96 @@ spv::LoopControlMask SPIRVEmitter::translateLoopAttribute(const Attr &attr) {
...
+  // Branch from the current insert point to the header block.
+  theBuilder.createBranch(headerBB);
+  theBuilder.addSuccessor(headerBB);
+
+  // Process the <header> block
+  // The header block must always branch to the body.
+  theBuilder.setInsertPoint(headerBB);
+  theBuilder.createBranch(bodyBB, mergeBB, continueBB, loopControl);
+  theBuilder.addSuccessor(bodyBB);
+  theBuilder.addSuccessor(mergeBB);
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: I think `<header>` only has `bodyBB` as its successor?

File: tools/clang/include/clang/SPIRV/ModuleBuilder.h (hunk 2 above)
Comment (ehsannas) [inline tools/clang/include/clang/SPIRV/ModuleBuilder.h]: Done

File: tools/clang/include/clang/SPIRV/ModuleBuilder.h (hunk 1 above)
Comment (ehsannas) [inline tools/clang/include/clang/SPIRV/ModuleBuilder.h]: ouch! I can't believe I didn't see that! :)
Done.

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp (hunk 3 above)
Comment (ehsannas) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp (hunk 4 above)
Comment (ehsannas) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done

**Pull request merged**